support KV transfer of MLA backend

chenqianfzh opened the pull request:
This PR support PD disaggregation for MLA, which is supported in deepseek models. This PR is aligned with the latest vllm of ver 0.8.3.

sample commands I was using:
```
VLLM_ATTENTION_BACKEND=FLASHMLA VLLM_USE_V1=0 LMCACHE_USE_EXPERIMENTAL=True  CUDA_VISIBLE_DEVICES=6  LMCACHE_CONFIG_FILE=/data/rain_dev/example.yaml python3 -m vllm.entrypoints.openai.api_server --model /data/models/DeepSeek-V2-Lite --port 7080 --max-num-seqs 1 --trust-remote-code --enforce-eager --gpu-memory-utilization 0.90 --swap-space 0 --enable-chunked-prefill=False --kv-transfer-config '{"kv_connector":"LMCacheConnector","kv_role":"kv_producer"}' 


VLLM_ATTENTION_BACKEND=FLASHMLA VLLM_USE_V1=0 LMCACHE_USE_EXPERIMENTAL=True  CUDA_VISIBLE_DEVICES=7  LMCACHE_CONFIG_FILE=/data/rain_dev/example.yaml python3 -m vllm.entrypoints.openai.api_server --model /data/models/DeepSeek-V2-Lite --port 7090 --max-num-seqs 1 --trust-remote-code --gpu-memory-utilization 0.90 --swap-space 0 --enable-chunked-prefill=False --kv-transfer-config '{"kv_connector":"LMCacheConnector","kv_role":"kv_consumer"}' 
```

maobaolong commented on lmcache/integration/vllm/vllm_adapter.py:
```diff
@@ -77,21 +85,38 @@ def init_lmcache_engine(
     head_size = model_config.get_head_size()
-    kv_shape = (num_layer, 2, chunk_size, num_kv_head, head_size)
+    if use_mla:
+        kv_shape = (num_layer, 1, chunk_size, 1, head_size)
```
The kvcache shape of mla is 3 dim without layer dim, which is [blockCount, headNum, headSize]

The following is the log I print for the one layer kvcache shape

```
recv [rank0]: is mla True, the kvcache[0].shape=torch.Size([23492, 64, 576])

```

YaoJiayi replied:
@maobaolong I think this is the shape of the cpu tensor (destination tensor). Looks good to me:)

maobaolong commented:
@chenqianfzh Compare to https://github.com/LMCache/LMCache/pull/368/  ,  I cannot find the place which store hidden states, is it not necessary to store?

YaoJiayi reviewed (commented):
@chenqianfzh The code looks good to me! Just one question: which do we have another separate class for MLA connector? Can't it reuse the existing GPU connector? Also, the format checker seems to be failing:)

chenqianfzh commented:
> @chenqianfzh Compare to #368 , I cannot find the place which store hidden states, is it not necessary to store?

per offline discussion with LMCache folks, this PR will be about MLA cache transfer only. So for now, the consumer will recompute everything.

chenqianfzh commented:
> @chenqianfzh The code looks good to me! Just one question: which do we have another separate class for MLA connector? Can't it reuse the existing GPU connector? Also, the format checker seems to be failing:)

Thanks for reviewing it. 

Actually,  started from revising the existing connector. However, I found it is gives better code readability and maintainability to create a new class due to the following difference in MLA architecture:

1. different KV cache shape, so different get_shape() func
2. different parameters to call the cuda code of multi_layer_kv_transfer
3. In the connector of MLA, we dont't need the member variable of page_buffer_size
4. In the MLA KV cache, the member of hidden_sim_size is not needed. Instead, we need aligned_head_size instead.

I have to put six or seven if/else in the Connector class if I reuse the existing class, which is prone to regression and hard to update. 

Besides, the code is formatted. 

Thanks.

chenqianfzh committed:
mla kv cache transfer

chenqianfzh committed:
running format.sh

YaoJiayi commented:
> > @chenqianfzh The code looks good to me! Just one question: which do we have another separate class for MLA connector? Can't it reuse the existing GPU connector? Also, the format checker seems to be failing:)
> 
> Thanks for reviewing it.
> 
> Actually, started from revising the existing connector. However, I found it is gives better code readability and maintainability to create a new class due to the following difference in MLA architecture:
> 
>     1. different KV cache shape, so different get_shape() func
> 
>     2. different parameters to call the cuda code of multi_layer_kv_transfer
> 
>     3. In the connector of MLA, we dont't need the member variable of page_buffer_size
> 
>     4. In the MLA KV cache, the member of hidden_sim_size is not needed. Instead, we need aligned_head_size instead.
> 
> 
> I have to put six or seven if/else in the Connector class if I reuse the existing class, which is prone to regression and hard to update.
> 
> Besides, the code is formatted.
> 
> Thanks.

Ok. Then we can use another separate class then. Btw, could you also run pytest locally as well? The unit tests are also failing. I think you have to make the `use_mla` argument default with `False`

thesues committed:
add new tests for MLA kv transfer

chenqianfzh commented:
@YaoJiayi I added new tests to cover MLA kv transfer. Also fixed the broken. 

Thanks.

YaoJiayi commented:
> @YaoJiayi I added new tests to cover MLA kv transfer. Also fixed the broken.
> 
> Thanks.

@chenqianfzh I think it's still failing? I got error messages like
`FAILED tests/benchmarks/test_benchmark.py::test_cachegen_decoder_bench[64-vllm] - TypeError: LMCacheEngineMetadata.__init__() missing 1 required positional argument: 'use_mla'`.

I think the `use_mla` argument should be optional?

chenqianfzh commented:
> > @YaoJiayi I added new tests to cover MLA kv transfer. Also fixed the broken.
> > Thanks.
> 
> @chenqianfzh I think it's still failing? I got error messages like `FAILED tests/benchmarks/test_benchmark.py::test_cachegen_decoder_bench[64-vllm] - TypeError: LMCacheEngineMetadata.__init__() missing 1 required positional argument: 'use_mla'`.
> 
> I think the `use_mla` argument should be optional?

sure, I will fix it.

I would like to see the unit test result in CI/CD. Yet I always got "page not found". 
https://buildkite.com/lmcache/lmcache-unittests/builds/925 is the link I just tried. What did I do wrong?

Thanks

thesues committed:
fix tests

YaoJiayi commented:
@chenqianfzh I think you are able to run tests locally with `pytest` under `tests` folder?

chenqianfzh commented:
yep. Finally it is all green now :-)

Shaoting-Feng commented:
Thanks for your PR. I ran some E2E performance tests on your branch and on the latest dev branch. The TTFT with CPU offloading has increased from 0.2s to 0.5s on one A100. It uses Flash Attention backend, so this performance degrade is pretty weird. Could you test with your environment? The scripts are as followed:

```bash
git clone https://github.com/LMCache/lmcache-tests
cd lmcache-tests
pip install -r requirements.txt
python3 main.py tests/tests.py -f test_local_cpu_experimental -o outputs/
cd outputs
python3 process_result.py (need to change the csv file name in line 61)
```

This should generate a pdf file. Thanks!

YaoJiayi commented:
> Thanks for your PR. I ran some E2E performance tests on your branch and on the latest dev branch. The TTFT with CPU offloading has increased from 0.2s to 0.5s on one A100. It uses Flash Attention backend, so this performance degrade is pretty weird. Could you test with your environment? The scripts are as followed:
> 
> ```shell
> git clone https://github.com/LMCache/lmcache-tests
> cd lmcache-tests
> pip install -r requirements.txt
> python3 main.py tests/tests.py -f test_local_cpu_experimental -o outputs/
> cd outputs
> python3 process_result.py (need to change the csv file name in line 61)
> ```
> 
> This should generate a pdf file. Thanks!

@chenqianfzh IIUC, the code changes shouldn't affect the the default code path (without MLA) at all?  I think the above numbers from @Shaoting-Feng is just the comparison between latest dev branch and this MLA branch (`bytedance-iaas:qina/mla-on-083`) without enabling MLA.

Could you share some benchmark numbers as well (should be very simple according to @Shaoting-Feng 's script)?

chenqianfzh commented:
> > Thanks for your PR. I ran some E2E performance tests on your branch and on the latest dev branch. The TTFT with CPU offloading has increased from 0.2s to 0.5s on one A100. It uses Flash Attention backend, so this performance degrade is pretty weird. Could you test with your environment? The scripts are as followed:
> > ```shell
> > git clone https://github.com/LMCache/lmcache-tests
> > cd lmcache-tests
> > pip install -r requirements.txt
> > python3 main.py tests/tests.py -f test_local_cpu_experimental -o outputs/
> > cd outputs
> > python3 process_result.py (need to change the csv file name in line 61)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This should generate a pdf file. Thanks!
> 
> @chenqianfzh IIUC, the code changes shouldn't affect the the default code path (without MLA) at all? I think the above numbers from @Shaoting-Feng is just the comparison between latest dev branch and this MLA branch (`bytedance-iaas:qina/mla-on-083`) without enabling MLA.
> 
> Could you share some benchmark numbers as well (should be very simple according to @Shaoting-Feng 's script)?

I am on vacation now. Will do it, and some analysis (if necessary) when I am back. 

sorry for the delay.

chunxiaozheng committed:
[bugfix] set context chunk param when use mla

chunxiaozheng commented on lmcache/integration/vllm/vllm_adapter.py:
```diff
@@ -656,6 +682,12 @@ def lmcache_retrieve_kv(
     assert len(lmc_num_computed_tokens_list) == seq_cnt
     assert len(num_computed_tokens_list) == seq_cnt
 
+    is_all_prefill = all(
```
hi, @chenqianfzh In our testing, there were performance issues when turning off/on MLA. After testing, we found that removing these lines restored the issue. What are the functions of these lines?

```
is_all_prefill = all(
    [status == RetrieveStatus.PREFILL for status in retrieve_status])

if is_all_prefill and num_request_not_found == 0:
    return model_input, False, None
```

YaoJiayi replied:
@chenqianfzh
What's the point of the following condition?
```python
if is_all_prefill and num_request_not_found == 0:
    return model_input, False, None
```

chenqianfzh replied:
@chunxiaozheng the PD disaggregation is only supported in prefill-only scenarios, no the chunked prefill scenarios. 
So this condition is to make sure all the sequences in the requests are prefills. 

@YaoJiayi If we have the hidden states retrieved, this condition means that everything is in a good state. This is the place to return all the retrieved info to vllm.
However, per offline discussion, hidden states are not retrieved in this PR. so this is just a place holder for the future.

chunxiaozheng replied:
> @chunxiaozheng the PD disaggregation is only supported in prefill-only scenarios, no the chunked prefill scenarios. So this condition is to make sure all the sequences in the requests are prefills.
> 
> @YaoJiayi If we have the hidden states retrieved, this condition means that everything is in a good state. This is the place to return all the retrieved info to vllm. However, per offline discussion, hidden states are not retrieved in this PR. so this is just a place holder for the future.

@chenqianfzh We found in our testing that this may cause performance rollback. Could you help us take a look?

chunxiaozheng commented:
hi, @chenqianfzh, @YaoJiayi @maobaolong  when chunked_refill and prefix_cache are turned off in VLLM and MLA is enabled, we encountered a bug. The following positions in the `vllm/attention/backends/mla/common.py` of VLLM are reporting errors
```
assert prefill_metadata.context_chunk_seq_tot is not None
assert prefill_metadata.context_chunk_cu_seq_lens is not None
assert prefill_metadata.context_chunk_starts is not None
assert prefill_metadata.context_chunk_max_seq_lens is not None
```
The reason for the bug is that if there is a hit in lmcache, the `context_chunk_ *` related parameters are not processed when  rebuilt `rebuilt_tttn_metadata` in `build_partial_prefill_input()`. [This PR](https://github.com/bytedance-iaas/LMCache/pull/8) fixes this issue, could you please help review it? Thanks!

chenqianfzh commented:
> > Thanks for your PR. I ran some E2E performance tests on your branch and on the latest dev branch. The TTFT with CPU offloading has increased from 0.2s to 0.5s on one A100. It uses Flash Attention backend, so this performance degrade is pretty weird. Could you test with your environment? The scripts are as followed:
> > ```shell
> > git clone https://github.com/LMCache/lmcache-tests
> > cd lmcache-tests
> > pip install -r requirements.txt
> > python3 main.py tests/tests.py -f test_local_cpu_experimental -o outputs/
> > cd outputs
> > python3 process_result.py (need to change the csv file name in line 61)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This should generate a pdf file. Thanks!
> 
> @chenqianfzh IIUC, the code changes shouldn't affect the the default code path (without MLA) at all? I think the above numbers from @Shaoting-Feng is just the comparison between latest dev branch and this MLA branch (`bytedance-iaas:qina/mla-on-083`) without enabling MLA.
> 
> Could you share some benchmark numbers as well (should be very simple according to @Shaoting-Feng 's script)?

Hi, I ran the test in a H20 GPU based on your instructions. I found not TTFT increment. The following are what is shown in the generated pdf. 

![image](https://github.com/user-attachments/assets/602424b7-8247-438c-8640-0abe2bfe5a61)

![image](https://github.com/user-attachments/assets/107d44f3-6cbe-4794-ad21-28c9788303b1)

![image](https://github.com/user-attachments/assets/eb53c018-22c1-4264-a30e-309a8de9eac0)

![image](https://github.com/user-attachments/assets/40244a15-cdf1-4859-8106-e52835bbc5ef)


My test config is:
1.  lmcache-tests is based on the latest version (branch tip 437a3a33f2a32c82d04ba5294dc62519b0563424)。 Yet I made a tiny change. In tests/trests.py, I changed the default gpu_memory_utilization from 0.8 to 0.4, so both vllms can run in the same GPU).
2. I rebased this PR to the latest lmcache branch
3. the backend vllm is upgraded to 0.8.4.

@Shaoting-Feng could you review the test results to see if there is anything irregular?

chenqianfzh commented:
> hi, @chenqianfzh, @YaoJiayi @maobaolong when chunked_refill and prefix_cache are turned off in VLLM and MLA is enabled, we encountered a bug. The following positions in the `vllm/attention/backends/mla/common.py` of VLLM are reporting errors
> 
> ```
> assert prefill_metadata.context_chunk_seq_tot is not None
> assert prefill_metadata.context_chunk_cu_seq_lens is not None
> assert prefill_metadata.context_chunk_starts is not None
> assert prefill_metadata.context_chunk_max_seq_lens is not None
> ```
> 
> The reason for the bug is that if there is a hit in lmcache, the `context_chunk_ *` related parameters are not processed when rebuilt `rebuilt_tttn_metadata` in `build_partial_prefill_input()`. [This PR](https://github.com/bytedance-iaas/LMCache/pull/8) fixes this issue, could you please help review it? Thanks!

Could you give me the complete scenario that you ran into this issue, as I could not repro it in my side:
1. the version of vllm
2. the command line of starting vllm
3. the version of lmcache and the config file

thanks!

Shaoting-Feng commented:
> @Shaoting-Feng could you review the test results to see if there is anything irregular?

This test is sending ten identical contexts for each length to the serving engine. So for engine 0 (with LMCache), the TTFT should drop after the first request. 

But current results look like the LMCache is not reusing the KV Cache stored in the main memory.

chunxiaozheng commented:
> > hi, @chenqianfzh, @YaoJiayi @maobaolong when chunked_refill and prefix_cache are turned off in VLLM and MLA is enabled, we encountered a bug. The following positions in the `vllm/attention/backends/mla/common.py` of VLLM are reporting errors
> > ```
> > assert prefill_metadata.context_chunk_seq_tot is not None
> > assert prefill_metadata.context_chunk_cu_seq_lens is not None
> > assert prefill_metadata.context_chunk_starts is not None
> > assert prefill_metadata.context_chunk_max_seq_lens is not None
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The reason for the bug is that if there is a hit in lmcache, the `context_chunk_ *` related parameters are not processed when rebuilt `rebuilt_tttn_metadata` in `build_partial_prefill_input()`. [This PR](https://github.com/bytedance-iaas/LMCache/pull/8) fixes this issue, could you please help review it? Thanks!
> 
> Could you give me the complete scenario that you ran into this issue, as I could not repro it in my side:
> 
> 1. the version of vllm
> 2. the command line of starting vllm
> 3. the version of lmcache and the config file
> 
> thanks!

Sure, my vllm version is `0.8.4`, my start command is as follows:
```
LMCACHE_USE_EXPERIMENTAL=True LMCACHE_TRACK_USAGE=false \
VLLM_MLA_DISABLE=0 VLLM_USE_V1=0 \
LMCACHE_CONFIG_FILE=/disc/data1/cpu.yaml \
vllm serve /disc/data1/deepseek/DeepSeek-R1-AWQ \
           --trust-remote-code \
           --served-model-name vllm_cpu_offload \
           --max-model-len 32768 \
           --max-seq-len-to-capture 10000 \
           --max-num-seqs 64 \
           --gpu-memory-utilization 0.9 \
           --host 0.0.0.0 \
           -tp 8 --kv-transfer-config '{"kv_connector":"LMCacheConnector","kv_role":"kv_both","kv_parallel_size":2}'
```
the concent of `cpu.yaml` is:
```
chunk_size: 256
local_device: "cpu"
local_cpu: True
max_local_cpu_size: 100
remote_serde: "naive"
pipelined_backend: False
```
In our internal dataset, this will result in the above error, you can refer to `_compute_prefill_context(...)`, which is in `vllm.attention.backends.mla.common.py` in VLLM, we can see here that to enable mla, we need to set the `context_chunk_ *` parameters.

maobaolong commented:
@chenqianfzh Thanks for working on this great feature, it is a significant feature for LMCache. Hope the following reproduce steps can help you to reproduce the issue motioned.(Assert failed and cannot reuse cache).

I can reproduce the assert issue by the following step

1. Start an vLLM 0.8.4 container
```shell
docker run -d -it --rm --privileged --entrypoint /bin/bash --network host --name poolv0-084-mla --shm-size 512g --gpus all   -v /:/disc    mirrors.tencent.com/dop/vllm:v0.8.4
```
2. Install the lmcache with mla.

[dist_lmcache-tq-mbl-test.tar.gz](https://github.com/user-attachments/files/19951299/dist_lmcache-tq-mbl-test.tar.gz)

The above whl file is built from last commit 407f00f5dd of your branch of this PR.

```shell
pip install lmcache-0.1.4-cp312-cp312-linux_x86_64.whl
pip install torch==2.6.0
```

3. do some small modification to make partition match possible.

Remove this following code within `vllm_adapter.py`
<img width="1108" alt="image" src="https://github.com/user-attachments/assets/eb8ca7d7-d54c-4796-b56c-cb468172e865" />

Added a log to ensure the rebuild logic entered.
![Uploading image.png…]()


4. Executed the following command to start an Vllm instance
```console
LMCACHE_USE_EXPERIMENTAL=True LMCACHE_TRACK_USAGE=false VLLM_MLA_DISABLE=0  VLLM_USE_V1=0 \
LMCACHE_CONFIG_FILE=./lmcache-test.yaml vllm serve /disc/data1/deepseek/DeepSeek-V2-Lite-Chat/     \
--trust-remote-code            --served-model-name vllm_cpu_offload            --max-model-len 32768        \
--max-seq-len-to-capture 10000            --max-num-seqs 64            --gpu-memory-utilization 0.9          \
--host 0.0.0.0            -tp 1       \
--kv-transfer-config '{"kv_connector":"LMCacheConnector","kv_role":"kv_both","kv_parallel_size":2}'
```

5. Send two curl request to reproduce the assert issue.
```
curl http://localhost:8000/v1/chat/completions     -H "Content-Type: application/json"     -d '{
    "model": "vllm_cpu_offload",
    "messages": [{"role": "user", "content": "hi hi hi hi hi hi hi hi hi hi 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41"}],                                    
    "max_tokens": 2,
    "temperature": 0
    }'



curl http://localhost:8000/v1/chat/completions     -H "Content-Type: application/json"     -d '{
    "model": "vllm_cpu_offload",
    "messages": [{"role": "user", "content": "hi hi hi hi hi hi hi hi hi hi 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53"}],
    "max_tokens": 2,
    "temperature": 0
    }'
```

6. The assert failed log in the vLLM console.
```
[2025-04-28 19:18:01,013] LMCache INFO: rebuild_input=ModelInputForGPUWithSamplingMetadata(input_tokens=tensor([  207,    18,    22,   207,    18,    23,   207,    18,    24,   207,
           19,    15,   207,    19,    16,   207,    19,    17,   207,    19,
           18,   207,    19,    19,   207,    19,    20,   207,    19,    21,
          207,    19,    22,   207,    19,    23,   207,    19,    24,   207,
           20,    15,   207,    20,    16,   207,    20,    17,   207,    20,
           18,   185,   185, 77398,    25], device='cuda:0'), input_positions=tensor([112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125,
        126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139,
        140, 141, 142, 143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153,
        154, 155, 156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166],
       device='cuda:0'), token_types=None, seq_lens=[167], query_lens=[55], lora_mapping=None, lora_requests=set(), attn_metadata=FlashMLAMetadata(num_prefills=1, num_prefill_tokens=55, num_decode_tokens=0, slot_mapping=tensor([112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125,
        126, 127,   0,   1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,
         12,  13,  14,  15,  16,  17,  18,  19,  20,  21,  22,  23,  24,  25,
         26,  27,  28,  29,  30,  31,  32,  33,  34,  35,  36,  37,  38],
       device='cuda:0'), multi_modal_placeholder_index_maps=None, enable_kv_scales_calculation=False, use_cuda_graph=False, input_positions=tensor([  0,   1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12,  13,
         14,  15,  16,  17,  18,  19,  20,  21,  22,  23,  24,  25,  26,  27,
         28,  29,  30,  31,  32,  33,  34,  35,  36,  37,  38,  39,  40,  41,
         42,  43,  44,  45,  46,  47,  48,  49,  50,  51,  52,  53,  54,  55,
         56,  57,  58,  59,  60,  61,  62,  63,  64,  65,  66,  67,  68,  69,
         70,  71,  72,  73,  74,  75,  76,  77,  78,  79,  80,  81,  82,  83,
         84,  85,  86,  87,  88,  89,  90,  91,  92,  93,  94,  95,  96,  97,
         98,  99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111,
        112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125,
        126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139,
        140, 141, 142, 143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153,
        154, 155, 156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166],
       device='cuda:0'), seq_lens=[167], seq_lens_tensor=tensor([167], device='cuda:0', dtype=torch.int32), max_prefill_seq_len=167, max_decode_seq_len=0, context_lens_tensor=tensor([112], device='cuda:0', dtype=torch.int32), block_tables=tensor([[2, 2, 2, 2, 1, 1, 1, 1, 0, 0, 0]], device='cuda:0', dtype=torch.int32), max_query_len=55, max_decode_query_len=1, query_start_loc=tensor([ 0, 55], device='cuda:0', dtype=torch.int32), seq_start_loc=tensor([  0, 167], device='cuda:0', dtype=torch.int32), _cached_prefill_metadata=None, _cached_decode_metadata=None, head_dim=576, is_profile_run=False, context_chunk_cu_seq_lens=None, context_chunk_starts=None, context_chunk_seq_tot=None, context_chunk_max_seq_lens=None, context_chunk_workspace=None, decode_tile_scheduler_metadata=None, decode_num_splits=None), prompt_adapter_mapping=None, prompt_adapter_requests=set(), multi_modal_kwargs={}, request_ids_to_seq_ids={'chatcmpl-1f3ff1fa003a4775a0356c8a6fc821db': [1]}, finished_requests_ids=['chatcmpl-f0937b17fc704e7391b9b82af962910e'], virtual_engine=0, async_callback=functools.partial(<function weak_bind.<locals>.weak_bound at 0x7ffff7a47920>, ctx=<vllm.engine.llm_engine.SchedulerContext object at 0x7ffc1c1d6570>), scheduler_outputs=None, previous_hidden_states=None, sampling_metadata=SamplingMetadata(seq_groups=[SequenceGroupToSample(seq_ids=[1], sampling_params=SamplingParams(n=1, presence_penalty=0.0, frequency_penalty=0.0, repetition_penalty=1.0, temperature=0.0, top_p=1.0, top_k=-1, min_p=0.0, seed=None, stop=[], stop_token_ids=[], bad_words=[], include_stop_str_in_output=False, ignore_eos=False, max_tokens=2, min_tokens=0, logprobs=None, prompt_logprobs=None, skip_special_tokens=True, spaces_between_special_tokens=True, truncate_prompt_tokens=None, guided_decoding=None, extra_args=None), seq_data={1: SequenceData(prompt_token_ids=array('l', [100000, 5726, 25, 2868, 2868, 2868, 2868, 2868, 2868, 2868, 2868, 2868, 2868, 207, 16, 207, 17, 207, 18, 207, 19, 207, 20, 207, 21, 207, 22, 207, 23, 207, 24, 207, 16, 15, 207, 16, 16, 207, 16, 17, 207, 16, 18, 207, 16, 19, 207, 16, 20, 207, 16, 21, 207, 16, 22, 207, 16, 23, 207, 16, 24, 207, 17, 15, 207, 17, 16, 207, 17, 17, 207, 17, 18, 207, 17, 19, 207, 17, 20, 207, 17, 21, 207, 17, 22, 207, 17, 23, 207, 17, 24, 207, 18, 15, 207, 18, 16, 207, 18, 17, 207, 18, 18, 207, 18, 19, 207, 18, 20, 207, 18, 21, 207, 18, 22, 207, 18, 23, 207, 18, 24, 207, 19, 15, 207, 19, 16, 207, 19, 17, 207, 19, 18, 207, 19, 19, 207, 19, 20, 207, 19, 21, 207, 19, 22, 207, 19, 23, 207, 19, 24, 207, 20, 15, 207, 20, 16, 207, 20, 17, 207, 20, 18, 185, 185, 77398, 25]), output_token_ids=(), cumulative_logprob=0.0, get_num_computed_tokens=0)}, seq_len=167, query_len=55, generator=None, is_prompt=True, prompt_logprob_indices=[], sample_indices=[0])], selected_token_indices=tensor([54], device='cuda:0'), categorized_sample_indices={<SamplingType.GREEDY: 0>: tensor([0], device='cuda:0', dtype=torch.int32), <SamplingType.RANDOM: 1>: tensor([], device='cuda:0', dtype=torch.int32), <SamplingType.RANDOM_SEED: 2>: tensor([], device='cuda:0', dtype=torch.int32)}), is_prompt=True) (vllm_adapter.py:867:lmcache.integration.vllm.vllm_adapter)
CRITICAL 04-28 19:18:01 [launcher.py:116] MQLLMEngine is already dead, terminating server process
INFO:     127.0.0.1:35784 - "POST /v1/chat/completions HTTP/1.1" 500 Internal Server Error
ERROR 04-28 19:18:01 [engine.py:160] AssertionError()
ERROR 04-28 19:18:01 [engine.py:160] Traceback (most recent call last):
ERROR 04-28 19:18:01 [engine.py:160]   File "/usr/local/lib/python3.12/dist-packages/vllm/engine/multiprocessing/engine.py", line 158, in start
ERROR 04-28 19:18:01 [engine.py:160]     self.run_engine_loop()
ERROR 04-28 19:18:01 [engine.py:160]   File "/usr/local/lib/python3.12/dist-packages/vllm/engine/multiprocessing/engine.py", line 221, in run_engine_loop
ERROR 04-28 19:18:01 [engine.py:160]     request_outputs = self.engine_step()
ERROR 04-28 19:18:01 [engine.py:160]                       ^^^^^^^^^^^^^^^^^^
ERROR 04-28 19:18:01 [engine.py:160]   File "/usr/local/lib/python3.12/dist-packages/vllm/engine/multiprocessing/engine.py", line 247, in engine_step
ERROR 04-28 19:18:01 [engine.py:160]     raise e
ERROR 04-28 19:18:01 [engine.py:160]   File "/usr/local/lib/python3.12/dist-packages/vllm/engine/multiprocessing/engine.py", line 230, in engine_step
ERROR 04-28 19:18:01 [engine.py:160]     return self.engine.step()
ERROR 04-28 19:18:01 [engine.py:160]            ^^^^^^^^^^^^^^^^^^
ERROR 04-28 19:18:01 [engine.py:160]   File "/usr/local/lib/python3.12/dist-packages/vllm/engine/llm_engine.py", line 1431, in step
ERROR 04-28 19:18:01 [engine.py:160]     outputs = self.model_executor.execute_model(
ERROR 04-28 19:18:01 [engine.py:160]               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ERROR 04-28 19:18:01 [engine.py:160]   File "/usr/local/lib/python3.12/dist-packages/vllm/executor/executor_base.py", line 140, in execute_model
ERROR 04-28 19:18:01 [engine.py:160]     output = self.collective_rpc("execute_model",
ERROR 04-28 19:18:01 [engine.py:160]              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ERROR 04-28 19:18:01 [engine.py:160]   File "/usr/local/lib/python3.12/dist-packages/vllm/executor/uniproc_executor.py", line 56, in collective_rpc
ERROR 04-28 19:18:01 [engine.py:160]     answer = run_method(self.driver_worker, method, args, kwargs)
ERROR 04-28 19:18:01 [engine.py:160]              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ERROR 04-28 19:18:01 [engine.py:160]   File "/usr/local/lib/python3.12/dist-packages/vllm/utils.py", line 2378, in run_method
ERROR 04-28 19:18:01 [engine.py:160]     return func(*args, **kwargs)
ERROR 04-28 19:18:01 [engine.py:160]            ^^^^^^^^^^^^^^^^^^^^^
ERROR 04-28 19:18:01 [engine.py:160]   File "/usr/local/lib/python3.12/dist-packages/vllm/worker/worker_base.py", line 420, in execute_model
ERROR 04-28 19:18:01 [engine.py:160]     output = self.model_runner.execute_model(
ERROR 04-28 19:18:01 [engine.py:160]              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ERROR 04-28 19:18:01 [engine.py:160]   File "/usr/local/lib/python3.12/dist-packages/torch/utils/_contextlib.py", line 116, in decorate_context
ERROR 04-28 19:18:01 [engine.py:160]     return func(*args, **kwargs)
ERROR 04-28 19:18:01 [engine.py:160]            ^^^^^^^^^^^^^^^^^^^^^
ERROR 04-28 19:18:01 [engine.py:160]   File "/usr/local/lib/python3.12/dist-packages/vllm/worker/model_runner.py", line 1770, in execute_model
ERROR 04-28 19:18:01 [engine.py:160]     hidden_or_intermediate_states = model_executable(
ERROR 04-28 19:18:01 [engine.py:160]                                     ^^^^^^^^^^^^^^^^^
ERROR 04-28 19:18:01 [engine.py:160]   File "/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py", line 1739, in _wrapped_call_impl
ERROR 04-28 19:18:01 [engine.py:160]     return self._call_impl(*args, **kwargs)
ERROR 04-28 19:18:01 [engine.py:160]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ERROR 04-28 19:18:01 [engine.py:160]   File "/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py", line 1750, in _call_impl
ERROR 04-28 19:18:01 [engine.py:160]     return forward_call(*args, **kwargs)
ERROR 04-28 19:18:01 [engine.py:160]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ERROR 04-28 19:18:01 [engine.py:160]   File "/usr/local/lib/python3.12/dist-packages/vllm/model_executor/models/deepseek_v2.py", line 703, in forward
ERROR 04-28 19:18:01 [engine.py:160]     hidden_states = self.model(input_ids, positions, intermediate_tensors,
ERROR 04-28 19:18:01 [engine.py:160]                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ERROR 04-28 19:18:01 [engine.py:160]   File "/usr/local/lib/python3.12/dist-packages/vllm/compilation/decorators.py", line 172, in __call__
ERROR 04-28 19:18:01 [engine.py:160]     return self.forward(*args, **kwargs)
ERROR 04-28 19:18:01 [engine.py:160]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ERROR 04-28 19:18:01 [engine.py:160]   File "/usr/local/lib/python3.12/dist-packages/vllm/model_executor/models/deepseek_v2.py", line 660, in forward
ERROR 04-28 19:18:01 [engine.py:160]     hidden_states, residual = layer(positions, hidden_states, residual)
ERROR 04-28 19:18:01 [engine.py:160]                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ERROR 04-28 19:18:01 [engine.py:160]   File "/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py", line 1739, in _wrapped_call_impl
ERROR 04-28 19:18:01 [engine.py:160]     return self._call_impl(*args, **kwargs)
ERROR 04-28 19:18:01 [engine.py:160]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ERROR 04-28 19:18:01 [engine.py:160]   File "/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py", line 1750, in _call_impl
ERROR 04-28 19:18:01 [engine.py:160]     return forward_call(*args, **kwargs)
ERROR 04-28 19:18:01 [engine.py:160]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ERROR 04-28 19:18:01 [engine.py:160]   File "/usr/local/lib/python3.12/dist-packages/vllm/model_executor/models/deepseek_v2.py", line 562, in forward
ERROR 04-28 19:18:01 [engine.py:160]     hidden_states = self.self_attn(
ERROR 04-28 19:18:01 [engine.py:160]                     ^^^^^^^^^^^^^^^
ERROR 04-28 19:18:01 [engine.py:160]   File "/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py", line 1739, in _wrapped_call_impl
ERROR 04-28 19:18:01 [engine.py:160]     return self._call_impl(*args, **kwargs)
ERROR 04-28 19:18:01 [engine.py:160]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ERROR 04-28 19:18:01 [engine.py:160]   File "/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py", line 1750, in _call_impl
ERROR 04-28 19:18:01 [engine.py:160]     return forward_call(*args, **kwargs)
ERROR 04-28 19:18:01 [engine.py:160]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ERROR 04-28 19:18:01 [engine.py:160]   File "/usr/local/lib/python3.12/dist-packages/vllm/model_executor/models/deepseek_v2.py", line 479, in forward
ERROR 04-28 19:18:01 [engine.py:160]     return self.mla_attn(hidden_states_or_q_c,
ERROR 04-28 19:18:01 [engine.py:160]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ERROR 04-28 19:18:01 [engine.py:160]   File "/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py", line 1739, in _wrapped_call_impl
ERROR 04-28 19:18:01 [engine.py:160]     return self._call_impl(*args, **kwargs)
ERROR 04-28 19:18:01 [engine.py:160]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ERROR 04-28 19:18:01 [engine.py:160]   File "/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py", line 1750, in _call_impl
ERROR 04-28 19:18:01 [engine.py:160]     return forward_call(*args, **kwargs)
ERROR 04-28 19:18:01 [engine.py:160]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ERROR 04-28 19:18:01 [engine.py:160]   File "/usr/local/lib/python3.12/dist-packages/vllm/attention/layer.py", line 229, in forward
ERROR 04-28 19:18:01 [engine.py:160]     return torch.ops.vllm.unified_attention(
ERROR 04-28 19:18:01 [engine.py:160]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ERROR 04-28 19:18:01 [engine.py:160]   File "/usr/local/lib/python3.12/dist-packages/torch/_ops.py", line 1123, in __call__
ERROR 04-28 19:18:01 [engine.py:160]     return self._op(*args, **(kwargs or {}))
ERROR 04-28 19:18:01 [engine.py:160]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ERROR 04-28 19:18:01 [engine.py:160]   File "/usr/local/lib/python3.12/dist-packages/vllm/attention/layer.py", line 342, in unified_attention
ERROR 04-28 19:18:01 [engine.py:160]     return self.impl.forward(self, query, key, value, kv_cache, attn_metadata)
ERROR 04-28 19:18:01 [engine.py:160]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ERROR 04-28 19:18:01 [engine.py:160]   File "/usr/local/lib/python3.12/dist-packages/vllm/attention/backends/mla/common.py", line 1412, in forward
ERROR 04-28 19:18:01 [engine.py:160]     output[:num_prefill_tokens] = self._forward_prefill(
ERROR 04-28 19:18:01 [engine.py:160]                                   ^^^^^^^^^^^^^^^^^^^^^^
ERROR 04-28 19:18:01 [engine.py:160]   File "/usr/local/lib/python3.12/dist-packages/vllm/attention/backends/mla/common.py", line 1307, in _forward_prefill
ERROR 04-28 19:18:01 [engine.py:160]     context_output, context_lse = self._compute_prefill_context( \
ERROR 04-28 19:18:01 [engine.py:160]                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ERROR 04-28 19:18:01 [engine.py:160]   File "/usr/local/lib/python3.12/dist-packages/vllm/attention/backends/mla/common.py", line 1139, in _compute_prefill_context
ERROR 04-28 19:18:01 [engine.py:160]     assert prefill_metadata.context_chunk_seq_tot is not None
ERROR 04-28 19:18:01 [engine.py:160]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ERROR 04-28 19:18:01 [engine.py:160] AssertionError
```

rainj-me committed:
Merge pull request #8 from chunxiaozheng/support-mla

[bugfix] set context chunk param when use mla

rainj-me committed:
Merge branch 'dev' into qina/mla-on-083

rainj-me committed:
remove the all prefill check during cache rebuilt

chenqianfzh committed:
use the original model input if re-compute

YaoJiayi commented:
@chenqianfzh The format checker is still failing?

chenqianfzh committed:
running format.sh

chenqianfzh commented:
> @chenqianfzh The format checker is still failing?

Green now.

chunxiaozheng commented:
@chenqianfzh I found that `input_positions` is used in `FlashMLAMetadata`, so i try to rebuilt it in `build_partial_prefill_input`, the accuracy can increase from 0.22 to 0.32, therefore, I think that the issue with accuracy is still due to the `rebuilt_model_input` or `rebuilt_attn_metadata` not being constructed correctly.

chunxiaozheng committed:
[bugfix] fix mmlu accuracy issue

chenqianfzh committed:
Merge branch 'dev' into qina/mla-on-083

chenqianfzh committed:
fix num_prefill_tokens in rebuilt model_input

chunxiaozheng commented:
hi~ @maobaolong @chenqianfzh @YaoJiayi  could you please help review https://github.com/bytedance-iaas/LMCache/pull/9? Thanks! 
This PR fix the `mmlu` accuracy issue(https://github.com/LMCache/LMCache/issues/530), my test conditions are consistent with https://github.com/LMCache/LMCache/pull/428#issuecomment-2837160667, the result is as follows:
|  | first round accuracy | second round accuracy(full hit) | 
| ------ | ------ | ------ |
| enable MLA, and use `LMCacheConnector` | 0.879 | 0.879 |
| enable MLA, and do not use `LMCacheConnector` | 0.878 | 0.878 |

It can be seen that when using `LMCacheConnector` and use `mla`, there is no significant decrease in accuracy.

The CPU is set to `100GB`, in order to avoid eviction and ensure full hits in the second round, only the following files of `mmlu` were used in the test:
<img width="370" alt="image" src="https://github.com/user-attachments/assets/d2751829-f161-40d5-a05e-a2ec0e201adc" />

chenqianfzh committed:
Merge pull request #9 from chunxiaozheng/mla-accuracy-fix

[bugfix] fix mmlu accuracy issue

chenqianfzh committed:
Merge branch 'dev' into qina/mla-on-083

chenqianfzh committed:
fix after merge

maobaolong reviewed (approved):
As this has been tested by @chenqianfzh and @chunxiaozheng , it LGTM and thanks for the efforts and this significant feature!
@YaoJiayi Would you like to take a final check?

marquiz commented on lmcache/integration/vllm/vllm_adapter.py:
```diff
@@ -116,31 +152,44 @@ def init_lmcache_engine(
-            device=device)
+    if use_mla:
+        if use_layerwise:
+            raise ValueError("layereise MLA connector is not supported yet")
```
```suggestion
            raise ValueError("layerwise MLA connector is not supported yet")
```

chenqianfzh replied:
Thanks for catching.

Fixed.

chenqianfzh committed:
Merge branch 'dev' into qina/mla-on-083

chenqianfzh committed:
fix typo

YaoJiayi commented on docker/patch/lmcache_connector.py:
```diff
@@ -1,9 +1,10 @@
+# SPDX-License-Identifier: Apache-2.0
```
Changes in this folder are not needed anymore? Sincewe already have upstream support for both vllm v0 and v1.

chenqianfzh replied:
it is necessary, actually, as we changed the API of init_lmcache_engine(), adding scheduler_config.

I will update the code in the vllm side once this PR is merged.

YaoJiayi commented on lmcache/experimental/gpu_connector.py:
```diff
@@ -676,3 +676,151 @@ def batched_from_gpu(self, memory_objs: List[List[MemoryObj]],
+            memory_obj.tensor.copy_(tmp_gpu_buffer, non_blocking=True)
+
+        torch.cuda.synchronize()
+        memory_obj.metadata.fmt = MemoryFormat.KV_2LTD
```
Is the KV shape/format still `KV_2LTD` (i.e., [2, num layer, token length, hidden dim]) in MLA?

YaoJiayi replied:
Maybe consider adding a new KV format?

chenqianfzh replied:
updated as suggested, adding a new format.

YaoJiayi reviewed (approved):
LGTM! Left some minor comments

chenqianfzh committed:
Merge branch 'dev' into qina/mla-on-083

chenqianfzh committed:
update the KV format for MLA connector
